doc/computing/randomness.rst

```diff
@@ -0,0 +1,315 @@
+clones.
+
+If an int were passed, `a` and `b` would be exact clones and they would not
+influence each other.
+
+This is an important thing to remember because :func:`~sklearn.clone` is
+called everywhere in scikit-learn tools: in particular, most meta-estimators
+that accept non-fitted estimators will in fact call :func:`~sklearn.clone`
+internally (:class:`~sklearn.model_selection.GridSearchCV`,
+:class:`~sklearn.ensemble.StackingClassifier`,
+:class:`~sklearn.calibration.CalibratedClassifierCV`, etc.).
+
+CV splitters
+............
+
+When passed a `RandomState` instance, cv splitters yield different splits
+each time `split` is called. This can lead to dramatic mistakes when
+comparing the performance of different estimators::
+
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.ensemble import GradientBoostingClassifier
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.model_selection import KFold
+    >>> from sklearn.model_selection import cross_val_score
+    >>> import numpy as np
+
+    >>> rng = np.random.RandomState(0)
+    >>> X, y = make_classification(random_state=rng)
+    >>> rf = RandomForestClassifier(random_state=rng)
+    >>> gbdt = GradientBoostingClassifier(random_state=rng)
+    >>> cv = KFold(shuffle=True, random_state=rng)
+
+    >>> for est in (rf, gbdt):
+    ...     print(cross_val_score(est, X, y, cv=cv))
+    [0.85 0.95 0.9  0.95 0.95]
+    [0.85 0.7  0.95 0.8  0.85]
+
+Directly comparing the performance of the random forest vs the gradient
+boosting estimator on each fold would be a methodological mistake: **the
+splits on which the estimators are evaluated are different**. Indeed,
+:func:`~sklearn.model_selection.cross_val_score` will internally call
+`cv.split` on the same :class:`~sklearn.model_selection.KFold` instance,
+but the splits will be different each time. This is also true for any tool
+that performs model selection via cross-validation, e.g.
+:class:`~sklearn.model_selection.GridSearchCV` and
+:class:`~sklearn.model_selection.RandomizedSearchCV`: scores are not
+comparable fold-to-fold across different calls to `search.fit`, since
+`cv.split` would have been called multiple times. Within a single call to
+`search.fit`, however, fold-to-fold comparison is possible since the search
+estimator only calls `cv.split` once.
+
+For comparable fold-to-fold results in all scenarios, one should pass an int
+to the CV plitter: `KFold(shuffle=True, random_state=0)`.
+
+.. note::
+    What matters in this example is what was passed to
+    :class:`~sklearn.model_selection.KFold`. Whether we pass a `RandomState`
+    instance or an int to :func:`~sklearn.datasets.make_classification` or to
+    the estimators is not relevant for our illustration purpose. It does
+    however have an impact on the cross-validation procedure as explained
+    above, but it isn't what makes the fold-to-fold comparison incorrect.
```

Comment:
As above, I find this note more 'distracting' than helpful to understand. `make_classification` can be called without `random_state` and the two estimators in a deterministic way, then the note becomes unnecessary.